### PR TITLE
added substreams-sink-rs

### PR DESCRIPTION
**Pull request by vadim-frolov**
A basic Rust streaming consumer for Substreams.

### CODE REVIEW

File: substreams-sink-rs/proto/sf/substreams/v1/modules.proto [hunk 1]
```diff
@@ -0,0 +1,94 @@
+syntax = "proto3";
```
Comment (benjaminbollen) [inline substreams-sink-rs/proto/sf/substreams/v1/modules.proto:1]: do we need to hard-copy these streaming fast proto files or can we depend on them through reference?

File: substreams-sink-rs/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                .map_err(|_| Status::invalid_argument("failed to read package".to_string()))?,
+        )
+        .map_err(|_| Status::invalid_argument("failed to decode package".to_string()))?;
+
+        let modules = pkg.modules.ok_or(Status::invalid_argument(
+            "failed to find modules in package".to_string(),
+        ))?;
+
+        let request = Request {
+            start_block_num,
```
Comment (benjaminbollen) [inline substreams-sink-rs/src/lib.rs:68]: curious if later in life we can type this stronger, but unsure whether that buys us anything

File: substreams-sink-rs/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            output_modules: vec![module_name.to_string()],
+            production_mode: true,
+            debug_initial_store_snapshot_for_modules: vec![],
+            modules: Some(modules),
+            output_module: module_name.to_string(),
+        };
+        self.blocks(request).await
+    }
+
+    pub fn send_compressed(mut self) -> Self {
```
Comment (benjaminbollen) [inline substreams-sink-rs/src/lib.rs:82]: good tasks would be to document the public functions

Comment (benjaminbollen) [reply substreams-sink-rs/src/lib.rs:82]: these are remapping the `tonic` grpc functions right?

File: substreams-sink-rs/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        self.inner = self.inner.send_compressed(CompressionEncoding::Gzip);
+        self
+    }
+
+    pub fn accept_compressed(mut self) -> Self {
+        self.inner = self.inner.accept_compressed(CompressionEncoding::Gzip);
+        self
+    }
+    pub async fn blocks(
```
Comment (benjaminbollen) [inline substreams-sink-rs/src/lib.rs:91]: what s the intended use of `blocks()`?

File: substreams-sink-rs/README.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+            &package_file_name,
+            &module_name,
+            start_block,
+            end_block,
+            "",
+            "STEP_IRREVERSIBLE",
+        )
+        .await
+        .unwrap()
+        .into_inner();
```
Comment (benjaminbollen) [inline substreams-sink-rs/README.md:31]: this feels a bit like we're breaking the interface contract? we're wrapping around tonic, but to get to get to the iterator we have to use tonic explicitly?

**Review by benjaminbollen — COMMENTED** [review]
very good; left some questions

I ll push a small commit with some small changes I saw while reading; will let you review that separately

File: substreams-sink-rs/proto/sf/substreams/v1/modules.proto (hunk 1 above)
Comment (vadim-frolov) [reply substreams-sink-rs/proto/sf/substreams/v1/modules.proto:1]: resolved in PR #18

File: substreams-sink-rs/src/lib.rs (hunk 2 above)
Comment (vadim-frolov) [reply substreams-sink-rs/src/lib.rs:68]: thanks for sharing your concern. will keep an eye on it

File: substreams-sink-rs/src/lib.rs (hunk 3 above)
Comment (vadim-frolov) [reply substreams-sink-rs/src/lib.rs:82]: simplified in PR #18

File: substreams-sink-rs/src/lib.rs (hunk 4 above)
Comment (vadim-frolov) [reply substreams-sink-rs/src/lib.rs:91]: simplified in #18

File: substreams-sink-rs/README.md (hunk 5 above)
Comment (vadim-frolov) [reply substreams-sink-rs/README.md:31]: made it more justified through simplification in PR #18

**Pull request merged**